packages/waku/src/adapters/vercel.ts

```diff
@@ -0,0 +1,139 @@
+import path from 'node:path';
+import {
+  rmSync,
+  cpSync,
+  existsSync,
+  mkdirSync,
+  readFileSync,
+  writeFileSync,
+} from 'node:fs';
+import { Hono } from 'hono';
+import { getRequestListener } from '@hono/node-server';
+import type { MiddlewareHandler } from 'hono';
+import {
+  unstable_createServerEntryAdapter as createServerEntryAdapter,
+  unstable_constants as constants,
+  unstable_honoMiddleware as honoMiddleware,
+} from 'waku/internals';
+
+const { DIST_PUBLIC, DIST_ASSETS } = constants;
+const { contextMiddleware, rscMiddleware, middlewareRunner } = honoMiddleware;
+
+export const vercelAdapter = createServerEntryAdapter(
+  (
+    { processRequest, processBuild, setAllEnv, config },
+    options?: {
+      static?: boolean;
+      middlewareFns?: (() => MiddlewareHandler)[];
+      middlewareModules?: Record<
+        string,
+        () => Promise<{
+          default: () => MiddlewareHandler;
+        }>
+      >;
+    },
+  ) => {
+    const { middlewareFns = [], middlewareModules = {} } = options || {};
+    setAllEnv(process.env as any);
+    const app = new Hono();
+    app.use(contextMiddleware());
+    for (const middlewareFn of middlewareFns) {
+      app.use(middlewareFn());
+    }
+    app.use(middlewareRunner(middlewareModules));
+    app.use(rscMiddleware({ processRequest }));
+    app.notFound((c) => {
+      const file = path.join(config.distDir, DIST_PUBLIC, '404.html');
+      if (existsSync(file)) {
+        return c.html(readFileSync(file, 'utf8'), 404);
+      }
+      return c.text('404 Not Found', 404);
+    });
+    return {
+      fetch: app.fetch,
+      build: processBuild,
+      postBuild: () => buildVercel({ ...config, serverless: !options?.static }),
+      listener: getRequestListener(app.fetch),
+    };
+  },
+);
+
+async function buildVercel({
```

Comment:
I don't think we should mix "build" logic and "runtime" logic. Previously `buildVercel` runs as plugin logic, but this now it's bundled in server production output.

For example, this wouldn't be an issue if we would integrate `@vercel/nft` (see comments below) later on since that dependency is a huge build time utility, not meant for server runtime.

Also the issue would manifest for cloudflare.

Reply:
I think I understand what you mean, but we are already mixing the runtime logic and build logic with `handleRequest` and `handleBuild`. Only what I can think of is `DO_NOT_BUNDLE` hack, which doesn't work for relative path...

Do you have any other idea? Maybe we provide the build logic as string???

Reply:
> I think I understand what you mean, but we are already mixing the runtime logic and build logic with `handleRequest` and `handleBuild`

Right, it's a gray area, but technically `processBuild/handleBuild` implementation itself didn't do file writing, but only passes render result to `emitFileInTask` as API. I thought that's an intentional API design. For example, maybe in far future, `emitFileInTask` could be implemented as RPC and if `processBuild/handleBuild` running inside cloudflare runtime calls it, the actual file writing can happen on node / plugin side runtime.

> Do you have any other idea? Maybe we provide the build logic as string???

I haven't follow the idea about new adapter system and don't have answer yet.

Reply:
Good point. Originally, I thought about `emiFile` thingy for the deploy adapter. Let me see if that works...

Reply:
> Let me see if that works...

It's probably unlikely.

> I think it should be `if (import.meta.BUILD) { ... }` or something. Should it work?

Trying this.


@hi-ogawa Can you spot what's wrong with 220ef87 ?

Reply:
My suggesting is to avoid relying on tree-shaking. The code there is obvious server runtime logic and obvious build orchestration logic, so it just feels wrong to me.

Reply:
https://github.com/wakujs/waku/pull/1707#discussion_r2404966955

Hm, that might be acceptable if the "internal" comes from a different entry point.
Like, `waku/adapters/vercel` and `waku/adapters/vercel/internal`.

> I cannot help if the goal is mixing.

The goal is actually not "mixing". The goal is that the adapter can be provided as a third-party library.

Reply:
> I'm assuming that the direction is to include adapter choice in server entry itself, but personally I don't think that's a good idea.

I'd like to have a single entry file as opposed to the proposal in #1388.
waku.config.ts or CLI option are other ideas, but I liked the current direction until I know the limitation (but I still think `/internal` entry point workaround is valid.)

Before deciding the direction to go, I want to have @rmarscher in the loop.